<html>
    <head>
        <title>VCF</title>
        
        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min.css">
        <script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap.native/2.0.27/bootstrap-native-v4.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/pako/1.0.11/pako.min.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/jszip/3.5.0/jszip.min.js"></script>
    
        <!-- Bootstrap CSS -->
        <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
        
        <link rel="icon" href="favicon.ico" type="image/x-icon"/>
        <script src="vcf.js"></script>
        <link rel="stylesheet" href="vcf.css">
        
        <script src="gendoc.js"></script>
        <link rel="stylesheet" href="gendoc.css">
        
        <link href="https://cdn.datatables.net/1.13.1/css/jquery.dataTables.min.css" rel="stylesheet"  crossorigin="anonymous">
        
    </head>
    
    <body>
        <div class="container-fluid">
            <div class="container">
            
                <section id="header">
                    <h1>
                        VCF 
                        <sup style="font-size:medium;color:green"><a href="https://dceg.cancer.gov/" target="_blank" style="color:green">DCEG</a></sup> <br>
                        <span style="font-size:small;color:blue">
                            [<a href="https://github.com/episphere/vcf" target="_blank">code</a>] 
                            [<a href="https://github.com/episphere/vcf/issues" target="_blank">issues</a>] 
                            [<a href="https://observablehq.com/@episphere/vcf" target="_blank" style="font-size:small;">notebook</a>] 
                            [<a href="https://gitter.im/episphere/terra" target="_blank">gitter</a>]
                        </span>
                    </h1>
                    
                    <hr>
                </section>
            
                <ul class="nav nav-tabs" id="options" role="tablist">
                  <li class="nav-item" role="presentation">
                    <button class="nav-link active" id="home-tab" data-bs-toggle="tab" data-bs-target="#home" type="button" role="tab" aria-controls="home" aria-selected="true">About VCF Library</button>
                  </li>
                  
                  <li class="nav-item" role="presentation">
                    <button class="nav-link" id="documentation-tab" data-bs-toggle="tab" data-bs-target="#documentation" type="button" role="tab" aria-controls="documentation" aria-selected="false">Documentation</button>
                  </li>
                </ul>
                
                <div class="tab-content" id="optionsContent">
                    <div class="tab-pane fade show active" id="home" role="tabpanel" aria-labelledby="home-tab">
                
                        <section id="introduction" class="mt-3">
                            <p>
                                Exploring mechanisms to traverse compressed Variant Calling Formatted files remotely: 
                                <b>without downloading them, and without a tbi file with content indexes</b>. 
                                The rationale is the same advanced for digital pathology in <a href="https://www.jpathinformatics.org/article.asp?issn=2153-3539;year=2020;volume=11;issue=1;spage=29;epage=29;aulast=Bremer" target="_blank">imagebox2</a> by relying on range requests. For parsing full files see <a href="https://ibl.github.io/vcf" target="_blank">old project at IBL</a>.
                            </p>
                            
                            <p> 
                                Although the real action is in the console, this page will be used to test the JavaScript vcf SDK. To do so, select the URL of the remote VCF file (or use default test VCF file borrowed from <a href="https://www.nature.com/articles/s41598-019-49114-z" target="_blank">Tollefson 2019</a>), and then try different methods. 
                                For examples of the real-world challange see <a href="https://ftp.ncbi.nlm.nih.gov/pub/clinvar" target="_blank">clinVar</a>, <a href="https://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/" target="_blank">snp</a> or <a href="http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/" target="_blank">1000genomes</a>.
                            </p>
                        </section>
                        
                        <section id="usage">
                            <div class="row">
                                <div class="col-12">
                                    <h3>Data</h3>
                                    <p> The data below will be used to demonstrate the main functions of our library, it contains all SNPs found in the Human genome. Make sure that the CORS is enabled in the data server, otherwise is not possible accessing it. </p>
                                    <input id="vcfURL" size=50 value="https://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/00-All.vcf.gz"></p>
                                </div>
                            </div>
                            
                            <div class="row">
                                <div class="col-12">
                                    <h3> 1 - Reading a slice of the data <span style="font-size:small; color:green; display: none;" id='infoFile'>(Filesize: <span id="filesize" style="color:red">...</span>)</span> </h3>
                                    <p> Choose an interval to read, then an object of VCF library will be initialized and will return the corresponding data with the chosen byte range.  </p>
                                    
                                    <form>
                                        <div class="form-group col-6">
                                            <input type="number" class="form-control mr-3 fields"  id="rangeStart" size=10 min=0 value=0 />
                                            <div  class="mr-3 fields " style="text-align: center;" > up to </div>
                                            <input type="number" class="form-control mr-3 fields"  id="rangeEnd" min=100 size=10 value=1000 />
                                            <button type="button" class="btn btn-primary" id="readRange" onclick="readRangeFun()" >Read</button>
                                        </div>    
                                    </form>
                                    <textarea id="rangeTextArea" style="width:20em;color:lime;background-color:black"> > Choose range and click on Read</textarea>
                                </div>
                            </div>
                            
                            <div class="row mt-3">
                                <div class="col-12">
                                    <h3> 2 - Filtering chromosomes and positions <span style="font-size:small; color:green; display: none;" id='infoTime'>(Query execution time: <span id="runtime" style="color:red">...</span>)</span> </h3>
                                    <p> 
                                    Choose a specific chromosome and position to search across the vcf file. Our search algorithm uses an strategy to optimize the search, returning the respective results in miliseconds. 
                                    <br />
                                    Examples to try it out: chromosome 7 and position 151040280 (7,151040280), chromosome 10 and position 133421085 (10,133421085), chromosome 17 and position 7675353 (17,7675353).
                                    </p>
                                    
                                    <form>
                                        <div class="form-group col-8">
                                            <div  class="mr-3 fields" style="text-align: right;" id='chroms' > Chromosome: </div>
                                            <select class="form-control mr-3 fields" style=" width: 200px !important;" id="chrom" value=0 > <option value='0'>Select a chromosome </option> </select>
                                            <div  class="mr-3 fields" style="text-align: right;" > Position: </div>
                                            <input type="number" class="form-control mr-3 fields" style=" width: 150px !important;" id="position" min=1000 size=10 value=1000 />
                                            <button type="button" class="btn btn-primary mr-3" id="filter" onclick="filterFun()" disabled=true>Filter</button>
                                            <button type="button" class="btn btn-primary" id="exporting" onclick="exportFun()" disabled=true>Export</button>
                                        </div>    
                                    </form>
                                    
                                </div>
                            </div>
                            
                            <div class="row mt-3">
                                <div class="col-12">
                                    <h3> 3 - Filtering a list of chromosomes and positions <span style="font-size:small; color:green; display: none;" id='infoTimeb'>(Query execution time: <span id="runtimeb" style="color:red">...</span>)</span> </h3>
                                    <p> 
                                    Given a list of sublists containing a chromosome and a position ([["8","73458588"],["MT","11252"],["4","53814975"]]), it will validate the list according to the loaded chromosomes in the file and return the SNPs found for the specified query. 
                                    <br />
                                    In this example, it will read <a href='multiple_query.json' >this file</a> and use 445 pairs chromosome-position in the query.
                                    </p>
                                    
                                    <form>
                                        <div class="form-group col-6">
                                            <button type="button" class="btn btn-primary mr-3" id="filterb" onclick="filterBatchFun()" disabled=true>Filter</button>
                                            <button type="button" class="btn btn-primary" id="exportingb" onclick="exportFun()" disabled=true>Export</button>
                                        </div>    
                                    </form>
                                    
                                </div>
                            </div>
                            
                            <div class="row mt-3">
                                <div class="col-12">
                                    <table id="filteredSnps"> </table>
                                </div>
                            </div>
                            
                            
                        </section>
                    </div>
                    
                    <div class="tab-pane fade" id="documentation" role="tabpanel" aria-labelledby="documentation-tab">
                        <section id="documentation-content" class="mt-3">
                            
                        </section>
                    </div>
                </div>
                
                <script>
                    //fetch("00-All.vcf.gz.idx.json").then(x=>x.json().then(x=>{idx=x;readRange.disabled=false}))
                    var v = {}
                    var datatable = {}
                    var columns = []
                    
                    readRangeFun=async _=>{
                        readRange.innerHTML="Reading ..."
                        readRange.disabled=true
                        infoFile.style.display="none"
                        
                        if(rangeTextArea.style.height.length==0){
                            rangeTextArea.style.height='10em'
                            rangeTextArea.style.width='100%'
                        }
                        
                        let url=vcfURL.value;   
                        let range=[parseInt(rangeStart.value),parseInt(rangeEnd.value)]
                        
                        v = await new Vcf(url)
                        var txtChrom = "<option value='0'>Select a chromosome </option> "
                        v.chrCode.forEach( x => { if(x!='0') { txtChrom+=`<option value="${x}">Chromosome ${x}</option>` } } )
                        chrom.innerHTML=txtChrom
                        
                        if( url.indexOf('.gz')!=-1 ){
                            let start=range[0]
                            
                            let res = await v.fetchGz(range)
                            rangeTextArea.value= (res.txt)
                        }else{
                            r = await v.fetchRange(range)
                            rangeTextArea.value=(await r.text())
                        }
                        var size = await vcf.fileSize(url)
                        filesize.innerHTML = (size/1000000).toFixed(2)+'M'
                        
                        readRange.innerHTML="Read"
                        readRange.disabled=false
                        infoFile.style.display="block"
                        
                        setTimeout(function(){
                            v.cols.forEach( x => { columns.push( {'title': x} ) } )
                            
                            $(document).ready(function() {
                                datatable = $('#filteredSnps').dataTable( {
                                    "data": [],
                                    "columns": columns
                                } );     
                                filteredSnps_wrapper.style.display='none'
                            } );
                            
                            filter.disabled=false
                            filterb.disabled=false
                        }, 3000)
                    }
                    
                    filterFun = async _=> {
                        filter.innerHTML="Filtering ..."
                        filter.disabled=true
                        filterb.disabled=true
                        exporting.disabled=true
                        
                        var chromosome = chrom.value
                        if(chromosome!='0'){
                            infoTime.style.display="none"
                            var start = performance.now()
                            
                            var pos = position.value
                            var q = `${chromosome},${pos}`
                            var result = await v.query(q)
                            filteredSnps_wrapper.style.display='block'
                            
                            var end = performance.now()
                            var diff=((end-start)/1000).toFixed(2)
                            runtime.innerHTML=  `${diff} seconds`
                            infoTime.style.display="block"
                            
                            $(document).ready(function() {
                               datatable.fnDestroy()
                               datatable = $('#filteredSnps').dataTable( {
                                    "data": result.hit,
                                    "columns": columns
                                } );     
                            } );
                        }
                        else{
                            alert('Select a valid chromosome')
                        }

                        filter.innerHTML="Filter"
                        filter.disabled=false
                        filterb.disabled=false
                        exporting.disabled=false
                    }
                    
                    filterBatchFun = async () => {
                        filterb.innerHTML="Filtering ..."
                        filterb.disabled=true
                        filter.disabled=true
                        exporting.disabled=true
                        
                        infoTimeb.style.display="none"
                        var start = performance.now()
                        
                        var dat = await fetch(location.href.split('#')[0]+'multiple_query.json')
                        dat = await dat.json()
                        var result = await v.queryInBatch( dat['list'] )    
                        filteredSnps_wrapper.style.display='block'
                        
                        var end = performance.now()
                        var diff=((end-start)/1000).toFixed(2)
                        runtimeb.innerHTML=  `${diff} seconds`
                        infoTimeb.style.display="block"
                        
                        $(document).ready(function() {
                           datatable.fnDestroy()
                           datatable = $('#filteredSnps').dataTable( {
                                "data": result.hit,
                                "columns": columns
                            } );     
                        } );
                        
                        filterb.innerHTML="Filter"
                        filterb.disabled=false
                        filter.disabled=false
                        exportingb.disabled=false
                    }
                    
                    exportFun = () => {
                        v.saveQueryResult()
                    }
                    
                    /*
                        regular expressions to get the list of query from idxx
                        rs(.*)\n - \n
                        (\d*)\n: \n(.*)\[ -> \n[
                        , '\n\n -> ],\n
                    */
                    
                    gendoc.getLines(location.href.split('#')[0]+'vcf.js').then( (value) => {
                        gendoc.buildDocumentation('documentation-content','https://github.com/episphere/vcf/blob/main/vcf.js')
                    })
                </script>
            </div>
        </div>

        <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.bundle.min.js" integrity="sha384-MrcW6ZMFYlzcLA8Nl+NtUVF0sA7MsXsP1UyJoMp4YLEuNSfAP+JcXn/tWtIaxVXM" crossorigin="anonymous"></script>
        <script src="https://code.jquery.com/jquery-3.5.1.js" crossorigin="anonymous"></script>
        <script src="https://cdn.datatables.net/1.13.1/js/jquery.dataTables.min.js" crossorigin="anonymous"></script>
        
    </body>
</html>

